### PR TITLE
[2.11.4.1] Fix compatibility with jackson 2.9.x

### LINKED ISSUE #451: ScalaObjectMapper backward compatibility
Hello, 

I have a library which has a dependency on `jackson-module-scala` version `2.9.7`. It contains this object: 

```
import com.fasterxml.jackson.databind._
import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper

val Mapper = new ObjectMapper() with ScalaObjectMapper
```

I excluded this dependency from the library and my project uses `jackson-module-scala` version `2.10.3`. However, it seems like this version was intended to support backward compatibility since it still has a deprecated `ScalaObjectMapper` defined in the `experimental` module. 

But I get this exception: 
```
Exception in thread "main" java.lang.NoSuchMethodError: com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper.$init$(Lcom/fasterxml/jackson/module/scala/experimental/ScalaObjectMapper;)V
```

So I'm not sure if the backward compatibility was done correctly. Do you think this should be fixed in version 2.10.x?

### PR DESCRIPTION
**Pull request by eugeniyk**
Fixes #451 for the cases when you have mix of old clients with 2.9.x and want to onboard gradually to 2.11.x

## Root causes

1. Missing $init$ in deprecated interfaces

decompiled java for 2.9.10 `ScalaObjectMapper.$init$` - under `com.fasterxml.jackson.module.scala.experimental` package:
```java
   static void $init$(final ScalaObjectMapper $this) {
      $this.com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$_setter_$com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$$MAP_$eq(Map.class);
      $this.com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$_setter_$com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$$OPTION_$eq(Option.class);
      $this.com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$_setter_$com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$$ITERABLE_$eq(Iterable.class);
   }
```

decompiled java for 2.11.4 `ScalaObjectMapper.$init$` - same code but under `com.fasterxml.jackson.module.scala` package.

same situation with `RequiredPropertiesSchemaModule`

2. Missing default interface methods in deprecated interfaces. Scala encoding in 2.12 uses default methods for trait methods 

before:
```java
public interface ScalaObjectMapper extends com.fasterxml.jackson.module.scala.ScalaObjectMapper
```

after:
```java
public interface ScalaObjectMapper extends com.fasterxml.jackson.module.scala.ScalaObjectMapper {
   ...
      // $FF: synthetic method
   Object com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$$super$readValue(final String content, final Manifest evidence$13);
   
   // $FF: synthetic method
   static Object readValue$(final ScalaObjectMapper $this, final String content, final Manifest evidence$8) {
      return $this.readValue(content, evidence$8);
   }

   default Object readValue(final String content, final Manifest evidence$8) {
      return this.com$fasterxml$jackson$module$scala$experimental$ScalaObjectMapper$$super$readValue(content, evidence$8);
   }
  ...
}
```

## Solution
1. Trick compiler to create `$init$` method by inserting custom code inside trait to emulate constructor behavior (be aware it shouldn't have any members, otherwise you'll have AbstractMember error).
2. override current members from stable interface inside deprecated one

Next time we can introduce [MIMA](https://github.com/lightbend/mima) to spot binary-compat issues

### CODE REVIEW

**Comment by pjfanning** [comment]
2.11 is no longer supported. I see no need to merge this.

File: build.sbt [hunk 1]
```diff
@@ -60,4 +60,4 @@ site.includeScaladoc()
 
 ghpages.settings
 
-git.remoteRepo := "git@github.com:FasterXML/jackson-module-scala.git"
+git.remoteRepo := "git@github.com:FasterXML/jackson-module-scala.git"
```
Comment (chameleon82) [inline build.sbt:63]: Please put back EOL

**Comment by eugeniyk** [comment]
> 2.11 is no longer supported. I see no need to merge this.

https://github.com/FasterXML/jackson#active-developed-jackson-2x-branches - 2.11 is still supported for patch releases, I wouldn't close branch for changes

@pjfanning the value I see is to provide safe ground between 2.9.x (which is 2017-2019 Dec, I believe still in use in many projects for these 3 years) and 2.11.x; 

I see many libraries has jackson as compile/runtime dependencies - which is fine, 2.9 vs 2.11/12 are binary compatible in java world as it should - semantic versioning is obeyed and you can easily substitute 1 version with another

Situation is different with scala module component and frankly it's sad situation for projects with 10+ dependencies all stuck on 2.9.x. _It is a workaround_ for such cases, and as you says if 2.11.x is not maintained anymore - let it be as the bridge to 2.12.x world for such projects.

> This commit would break binary compatibility for users coming from Jackson 2.10. There is no pretty solution to this problem this late in the day. If we had fixed the binary compatibility shortly after it was introduced then it would have been ok but at this late stage it is too late.

How this will break compat for users coming from `Jackson 2.10.x`? `experimental` package was originated in `2.9.x` and existed till `2.11.x`, were there any interface removed in 2.11?

**Comment by eugeniyk** [comment]
Check maven usages of jackson module scala: https://mvnrepository.com/artifact/com.fasterxml.jackson.module/jackson-module-scala

2.9.x has top place by [amount of users](https://mvnrepository.com/artifact/com.fasterxml.jackson.module/jackson-module-scala/2.9.9/usages) - this can let big projects evolve to 2.11.x without worrying about other stale components

Anyway, it will be maintainers' decision to let this change to be merged, at the end 1.5 year old library in an enterprise is literally up-to-date =)

**Comment by pjfanning** [comment]
I am not inclined to merge this. It is too late to backtrack. Feel free to fork the solution and to publish your own jar with this change.

**Comment by cowtowncoder** [comment]
I'll let @pjfanning make ultimate decision here, and will only say that further full patch releases of 2.10 are less likely than 2.11 (although not ruled out, unlike with 2.9). Likelihood of full 2.11 release depends on whether there are critical fixes for it -- I am guessing there's 50/50 chance for just one more patch release.

**Pull request closed without merging**